### PR TITLE
Reset offsets supervisor API

### PR DESCRIPTION
**Pull request by abhishekrb19**
This PR adds a new API: `/druid/indexer/v1/supervisor/<supervisorId>/resetOffsets`. This API accepts a reset datasource metadata in the body and only resets the specified partition offsets in the metadata store. Unlike the `/reset` API, this new API retains any previously checkpointed partition offsets not specified in the payload, so tasks will resume from where they left off. Further, only tasks serving the reset partitions will be restarted.

### Description

A few use cases:
- A supervisor that is reading from an offset that is invalid -- moved past the retention period, topic recreated, etc. So instead of a `/reset` that will clear the entire datasource metadata and start from the earliest or latest offset, users can now specify specific partition offsets.
- Users wanting some history for a streaming datasource, but not all the way to earliest can also benefit from this API.
- With support for multiple kafka streams by a single supervisor, a user can selectively reset a single topic's partition offsets instead of all the topics.

Why not re-purpose the existing `reset` API endpoint and the associated interface method?
 - To support backward compatibility with existing clients. The web-console (and possibly other clients) already send an empty body payload to the `reset` API, so accepting a new body in the existing API will cause some clients to break.

The API flow is as follows: 
- User submits the reset datasource metadata to the supervisor's `/resetOffsets` endpoint
- Few validations like the stream and metadata type are done before a `ResetOffsetsNotice` is created
- When the notice is handled asynchronously, the reset offsets is done, writing offsets to the metadata store and then restarting the appropriate tasks as necessary

#### Release note

Users can reset specific partition offsets for a supervisor using the `resetOffsets` API.

<hr>

##### Key changed/added classes in this PR
 * `SupervisorResource.java`
 * `Supervisor.java`
 * `SeekableStreamSupervisor.java`
 * `SeekableStreamSupervisorStateTest.java`
 * `TestSeekableStreamDataSourceMetadata.java`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaDataSourceMetadataTest.java [hunk 1]
```diff
@@ -170,4 +211,23 @@
...
+        GuiceInjectors.makeStartupInjector(),
+        Arrays.asList(
+            module,
+            (Module) binder -> {
+              binder.bindConstant().annotatedWith(Names.named("serviceName")).to("test");
+              binder.bindConstant().annotatedWith(Names.named("servicePort")).to(8000);
+              binder.bindConstant().annotatedWith(Names.named("tlsServicePort")).to(9000);
+            }
+        )
+    );
```
[the remark this answers is not in the document]
Comment (abhishekrb19) [reply extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaDataSourceMetadataTest.java]: Fixed

File: extensions-contrib/materialized-view-maintenance/src/main/java/org/apache/druid/indexing/materializedview/MaterializedViewSupervisor.java [hunk 2]
```diff
@@ -277,6 +277,12 @@ public void reset(DataSourceMetadata dataSourceMetadata)
     }
   }
 
+  @Override
+  public void resetOffsets(DataSourceMetadata resetDataSourceMetadata, boolean setOffsetsInMetadata)
```
Comment (AmatyaAvadhanula) [inline extensions-contrib/materialized-view-maintenance/src/main/java/org/apache/druid/indexing/materializedview/MaterializedViewSupervisor.java:281]: This method seems to be called with`setOffsetsInMetadata` always set to true

Comment (abhishekrb19) [reply extensions-contrib/materialized-view-maintenance/src/main/java/org/apache/druid/indexing/materializedview/MaterializedViewSupervisor.java]: yeah, thanks, cleaned up. This was a leftover from some refactoring. `ResetOffsets` always sets it to true and only `ResetNotice()` and `resetInternal()` have the additional parameter.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java [hunk 3]
```diff
@@ -998,10 +1006,66 @@ public void stop(boolean stopGracefully)
...
+        .getSeekableStreamSequenceNumbers()
+        .getPartitionSequenceNumberMap()
+        .entrySet()) {
+      final TaskGroup taskGroup = activelyReadingTaskGroups.get(getTaskGroupIdForPartition(resetPartitionOffset.getKey()));
+      if (taskGroup == null) {
+        throw InvalidInput.exception("No task found serving stream[%s], partition[%s] in supervisor[%s].",
+                                     resetStream, resetPartitionOffset.getKey(), supervisorId);
+      } else if (!taskGroup.startingSequences.containsKey(resetPartitionOffset.getKey())) {
+        throw InvalidInput.exception(
+            "Partition[%s] doesn't exist in checkpointed metadata for stream[%s] and supervisor[%s]."
```
Comment (jon-wei) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:1057]: Do we want to enforce that the partition must exist in the current checkpoint? (e.g., maybe someone repartitioned their stream and then wants to start from a specific offset in the newly added partitions)

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java [hunk 4]
```diff
@@ -998,10 +1006,66 @@ public void stop(boolean stopGracefully)
...
+      throw InvalidInput.exception(
+          "Datasource metadata instance does not match required, found instance of [%s].",
+          resetDataSourceMetadata.getClass()
+      );
+    }
+    @SuppressWarnings("unchecked")
+    final SeekableStreamDataSourceMetadata<PartitionIdType, SequenceOffsetType> resetMetadata =
+        (SeekableStreamDataSourceMetadata<PartitionIdType, SequenceOffsetType>) resetDataSourceMetadata;
+    String resetStream = resetMetadata.getSeekableStreamSequenceNumbers().getStream();
+    if (!resetMetadata.getSeekableStreamSequenceNumbers().getStream().equals(ioConfig.getStream())) {
```
[the remark this answers is not in the document]
Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java]: yep, done

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java (hunk 3 above)
Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java]: Good point, removed this check

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java [hunk 5]
```diff
@@ -3614,7 +3692,8 @@ private OrderedSequenceNumber<SequenceOffsetType> getOffsetFromStorageForPartiti
         if (!checkOffsetAvailability(partition, sequence)) {
           if (taskTuningConfig.isResetOffsetAutomatically()) {
             resetInternal(
-                createDataSourceMetaDataForReset(ioConfig.getStream(), ImmutableMap.of(partition, sequence))
+                createDataSourceMetaDataForReset(ioConfig.getStream(), ImmutableMap.of(partition, sequence)),
+                false
```
[the remark this answers is not in the document]
Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java]: I've pulled out the logic to its separate method

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java [hunk 6]
```diff
@@ -998,10 +1006,66 @@ public void stop(boolean stopGracefully)
...
+          resetStream,
+          supervisorId,
+          ioConfig.getStream()
+      );
+    }
+    for (Entry<PartitionIdType, SequenceOffsetType> resetPartitionOffset : resetMetadata
+        .getSeekableStreamSequenceNumbers()
+        .getPartitionSequenceNumberMap()
+        .entrySet()) {
+      final TaskGroup taskGroup = activelyReadingTaskGroups.get(getTaskGroupIdForPartition(resetPartitionOffset.getKey()));
```
[the remark this answers is not in the document]
Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java]: Adjusted based on our discussions, thanks! A user can issue a reset offsets even when the supervisor is suspended

**Comment by abhishekrb19** [comment]
The CodeQL alert about log entries depending on a user-provided value appears to be flagging existing logging code that's not part of this PR. I think it's a false positive alert.

File: docs/api-reference/supervisor-api.md [hunk 7]
```diff
@@ -3030,6 +3030,86 @@ Host: http://ROUTER_IP:ROUTER_PORT
...
+
+```HTTP
+POST /druid/indexer/v1/supervisor/social_media/resetOffsets HTTP/1.1
+Host: http://ROUTER_IP:ROUTER_PORT
+Content-Type: application/json
+
+{
+  "type": "kafka",
+  "partitions": {
+    "type": "end",
```
Comment (vogievetsky) [inline docs/api-reference/supervisor-api.md:3089]: What is `"type": "end",` ? where is it documented?

Comment (abhishekrb19) [reply docs/api-reference/supervisor-api.md:3104]: Added docs for the payload. Hope that clarifies it

**Review by AmatyaAvadhanula — APPROVED** [review]

**Comment by vogievetsky** [comment]
Ignoring the CodeQL issues per the comment above

**Pull request merged**